In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris
# Load dataset Iris dari scikit-learn
iris = load_iris()

### Import Library

Pada tahap ini dilakukan import library yang akan digunakan dalam proses klasifikasi Naïve Bayes.

- `numpy` digunakan untuk melakukan operasi pada array.
- `pandas` digunakan untuk mengolah data dalam bentuk DataFrame.
- `load_iris` digunakan untuk mengambil dataset Iris dari library `scikit-learn`.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris.target_names[iris.target]
# Definisikan X (fitur) dan y (target/label)
X = df.drop(columns=["Species"])
y = df["Species"]
print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


### Load Dataset Iris

Pada tahap ini dataset Iris dimuat menggunakan fungsi `load_iris()` dari `scikit-learn`.

Dataset kemudian diubah menjadi DataFrame menggunakan `pandas`. Selain empat fitur yang tersedia, ditambahkan kolom `Species` yang berisi jenis bunga Iris, yaitu `setosa`, `versicolor`, dan `virginica`.

Empat fitur yang digunakan adalah:
- Sepal Length
- Sepal Width
- Petal Length
- Petal Width

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

### Menentukan Fitur dan Target

Pada tahap ini dataset dibagi menjadi dua bagian, yaitu `X` sebagai fitur dan `y` sebagai target atau label.

Variabel `X` berisi empat fitur dari dataset Iris, sedangkan variabel `y` berisi species yang akan diprediksi.

Kolom `Species` dihapus dari `X` karena kolom tersebut merupakan target yang akan digunakan dalam proses klasifikasi.

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

### Label Encoding

Pada tahap ini label species yang masih berbentuk teks diubah menjadi bentuk angka menggunakan `LabelEncoder`.

Perubahan label yang dilakukan adalah:
- `setosa` menjadi `0`
- `versicolor` menjadi `1`
- `virginica` menjadi `2`

Pengubahan ini dilakukan agar label dapat digunakan dalam proses klasifikasi menggunakan model Naïve Bayes.

In [5]:
from sklearn.model_selection import train_test_split
X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

### Pembagian Data Training dan Testing

Pada tahap ini dataset dibagi menjadi data training dan data testing menggunakan `train_test_split()`.

Parameter `test_size=0.2` berarti 20% data digunakan sebagai data testing, sedangkan 80% sisanya digunakan sebagai data training.

Parameter `random_state=0` digunakan agar pembagian data tetap sama setiap kali program dijalankan.

In [ ]:
from sklearn.naive_bayes import GaussianNB
model = GaussianNB()
NB_model = model.fit(X_train, y_train)

### Membuat Model Gaussian Naïve Bayes

Pada tahap ini digunakan algoritma Gaussian Naïve Bayes untuk melakukan klasifikasi.

Gaussian Naïve Bayes digunakan karena fitur pada dataset Iris berupa data numerik kontinu.

Model kemudian dilatih menggunakan data training (`X_train` dan `y_train`) sehingga model dapat mempelajari hubungan antara fitur dengan species bunga Iris.

In [7]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

### Prediksi Data Testing

Pada tahap ini model yang telah dilatih digunakan untuk melakukan prediksi terhadap data testing (`X_test`).

Hasil prediksi disimpan pada variabel `result`. Hasil tersebut masih berupa angka yang mewakili masing-masing species, yaitu:

- `0` = setosa
- `1` = versicolor
- `2` = virginica

In [8]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


### Menghitung Accuracy

Pada tahap ini dilakukan evaluasi terhadap hasil prediksi menggunakan `accuracy_score()`.

Nilai accuracy diperoleh dengan membandingkan label sebenarnya pada data testing (`y_test`) dengan hasil prediksi model (`result`).

Berdasarkan hasil praktikum, diperoleh accuracy sebesar 96.67%. Hal ini menunjukkan bahwa model dapat mengklasifikasikan sebagian besar data testing dengan benar.

In [9]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\USER\AppData\Local\Programs\Python\Python313\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


### Menyiapkan Data Baru

Pada tahap ini disiapkan data baru yang akan diprediksi menggunakan model yang telah dibuat.

Data baru yang diberikan adalah:

`[7.3, 3.8, 4.6, 2.3]`

Nilai tersebut secara berurutan merupakan:
- Sepal Length = 7.3
- Sepal Width = 3.8
- Petal Length = 4.6
- Petal Width = 2.3

Data dibuat dalam bentuk DataFrame dan menggunakan nama kolom yang sama dengan dataset Iris agar struktur data sesuai dengan model yang telah dilatih.

In [10]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


### Prediksi Data Baru

Pada tahap ini data baru dimasukkan ke dalam model Gaussian Naïve Bayes menggunakan fungsi `predict()`.

Hasil prediksi awal berupa angka. Karena sebelumnya label species telah diubah menggunakan `LabelEncoder`, hasil tersebut dikembalikan ke nama species menggunakan `inverse_transform()`.

Berdasarkan hasil prediksi, data baru `[7.3, 3.8, 4.6, 2.3]` termasuk ke dalam species virginica.

*TUGAS*

In [20]:
# Prediksi inputan data baru

input_data = pd.DataFrame(
    np.array([[7.3, 3.8, 4.6, 2.3]]),
    columns=iris.feature_names
)

result = NB_model.predict(input_data)

prediction_label = lab.inverse_transform(result)

print(
    'Hasil prediksi untuk input baru adalah:',
    prediction_label
)

Hasil prediksi untuk input baru adalah: ['virginica']


### Prediksi Input Data Baru

Pada tahap ini dilakukan prediksi terhadap data baru menggunakan model Gaussian Naïve Bayes yang telah dilatih sebelumnya.

Data baru yang digunakan adalah `[7.3, 3.8, 4.6, 2.3]`. Data tersebut diubah menjadi DataFrame menggunakan nama kolom fitur dari dataset Iris agar memiliki struktur yang sama dengan data yang digunakan saat proses training.

Selanjutnya, `NB_model.predict()` digunakan untuk memprediksi species dari data baru tersebut. Hasil prediksi masih berupa label angka, sehingga digunakan `lab.inverse_transform()` untuk mengubah label tersebut kembali menjadi nama species.

Berdasarkan hasil prediksi, data baru `[7.3, 3.8, 4.6, 2.3]` diprediksi termasuk ke dalam species virginica.